nunaliit2-js/src/main/js/nunaliit2/n2.canvasVerticalTimeline.js

```diff
@@ -1,837 +1,974 @@
-			this._calcListItemWidth();
+			if (this.sortedElements.length > 0) {
+				timelineIndexOptions = {
+					'canvasId': this.canvasId,
+					'canvasIndexId': this.canvasIndexId,
+					'sortedElements': this.sortedElements,
+					'ascendingSortOrder': this.ascendingSortOrder
+				};
 
-			for (i = 0, e = this.sortedElements.length; i < e; i += 1) {
-				// Exclude link elements if produced by the generic element generator
-				if (this.elementGenerator._classname === 'GenericElementGenerator' 
-					&& this.sortedElements[i].isLink) {
-					// Do nothing
-				} else {
+				this.timelineIndex = new TimelineIndex(timelineIndexOptions);
+				this.sortedIndex = this.timelineIndex.getSortedIndex();
+				this.indexElements = this.timelineIndex.getIndexElements();
+				this._updateTimelinePadding();
+
+				for (i = 0, e = this.sortedElements.length; i < e; i += 1) {
 					timelineItemOptions = {
 						element: this.sortedElements[i],
-						timelineList: this.canvasListId,
-						itemWidth: this.itemWidth
+						canvasTimeLineId: this.canvasTimelineId,
 					};
 					new TimelineItem(timelineItemOptions);
 				}
+
+				this._linkIndexToListItems();
+				this.showService.fixElementAndChildren($timelineList);
 			}
+		},
+
+		_handleScrollEvent: function() {
+			var i, e, index, elem;
+			var headerHeight = 100;
+			if (this.sortedIndex && this.indexElements) {
+				for (i = 0, e = this.sortedIndex.length; i < e; i += 1) {
+					index = this.sortedIndex[i];
+					if (Object.hasOwnProperty.call(this.indexElements, index)
+						&& this.indexElements[index].id) {
 
-			this._linkIndexToListItems();
-			this.showService.fixElementAndChildren($timelineList);
-		}
-	},
-
-	_handleScrollEvent: function(){
-		var i, e, index, elem;
-		var headerHeight = 100;
-		if (this.sortedIndex && this.indexElements) {
-			for (i = 0, e = this.sortedIndex.length; i < e; i += 1){
-				index = this.sortedIndex[i];
-				if (this.indexElements.hasOwnProperty(index)) {
-					if (this.indexElements[index].id) {
 						elem = document.getElementById(this.indexElements[index].id);
-						if (elem && elem.getBoundingClientRect().top > headerHeight) {
+						if (elem
+							&& elem.getBoundingClientRect().top > headerHeight) {
 							this.timelineIndex.setActiveIndexItem(this.indexElements[index].label);
 							break;
 						}
 					}
 				}
 			}
-		}
-	},
+		},
 
-	_sortElements: function(){
-		var elementId, element, date, luxonDate;
-		var defaultDateFormat = "yyyy-LL-dd";
-		this.sortedElements = [];
+		// Sort elements based on sort value and label
+		_sortElements: function() {
+			var elementKeys, element, i, e, key;
+			var uniqueLabels = [];
+			this.sortedElements = [];
 
-		for (elementId in this.elementsById) {
-			if (this.elementsById.hasOwnProperty(elementId)) {
-				element = this.elementsById[elementId];
-				date = getDateFromDoc(element);
+			elementKeys = Object.keys(this.elementsById);
+			for (i = 0, e = elementKeys.length; i < e; i += 1) {
```

Comment:
I'm curious why we don't use a more succinct style.  I realise this approach is everywhere, and maybe it's a JS thing.  But couldn't the for loop condition be written like `for (var i = 0; i < elementKeys.length; i++)` ?  Enlighten me please!!